### PR TITLE
Path planner test suite

### LINKED ISSUE #770: Create A Generic Test Suite For All Path Planners
### Description of the task
Create a generic suite of tests to make sure we can verify that all path planners obey sane behavior. This will likely take the form of designing a series of scenarios that we expect all path planners to be able to handle and writing tests for them. This will allow us to more rapidly bring up new path planner algorithms in the future with much greater confidence.

### Acceptance criteria

- [ ] Implement a series of unit tests that represent the set of scenarios we expect all path planners to be able to handle.
- [ ] Test all path planners with the test suite, fix (or file tickets to fix) any bugs found

### Blocked By

#769 (and any tickets created from it)

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

Create a test suite for path planners

### Testing Done

Tests fail with straight line path planner

### Resolved Issues

Resolves #770 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/navigator/path_planner/path_planner_test.cpp [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        name_and_constructor<ThetaStarPathPlanner>(),
+        // uncomment this if you want some tests to fail
+        //        name_and_constructor<StraightLinePathPlanner>()
+};
+
+
+void validatePath(const Path &path, const Point &start, const Point &dest,
+                  const Rectangle &navigable_area, const std::vector<Obstacle> &obstacles)
+{
+    for (long i = 0; i < std::lround(1.0 / PATH_CHECK_INTERVAL); i++)
```
Comment (akhilveeraghanta) [inline src/software/ai/navigator/path_planner/path_planner_test.cpp]: Can we avoid `PATH_CHECK_INTERVAL`, by making it proportional to the speed/min obstacle size or other parameters like that to avoid potentially skipping over an obstacle?

My only concern is that if we gave a start position at the top left corner of the field, and a  dest position at the bottom right corner, (which we probably won't do irl), we would skip over smaller obstacles because the parameterization from 0-1 is not being scaled by the time or even distance

**Review by akhilveeraghanta — COMMENTED** [review]
Overall LG:tm:, super clean, just 1 comment

File: src/software/ai/navigator/path_planner/path_planner_test.cpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+                          {std::cos(4 * M_PI / 3), std::sin(4 * M_PI / 3)}),
+                      false),
+                  Obstacle::createRobotObstacle(
+                      TestUtil::createRobotAtPos(
+                          {std::cos(5 * M_PI / 3), std::sin(5 * M_PI / 3)}),
+                      false),
+              },
+          .should_return_path = true},
+
+         {.name           = "Circle of robots surrounding ego at distance 0.2",
```
Comment (garethellis0) [inline src/software/ai/navigator/path_planner/path_planner_test.cpp]: "ego" - I see someone has been working with autonomous cars in the valley...... :joy:

File: src/software/ai/navigator/path_planner/path_planner_test.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    else
+    {
+        EXPECT_FALSE(path.has_value())
+            << std::get<0>(GetParam()).first << " should not have returned path for "
+            << planner_test_case.name;
+    }
+}
+
+// uncomment this to run tests
+// cases still fail because theta* will intersect obstacles' bounding boxes at knots
```
Comment (garethellis0) [inline src/software/ai/navigator/path_planner/path_planner_test.cpp:288]: Can you elaborate on this a little bit / file an issue to fix this?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Really nice overall!

File: src/software/new_geom/rectangle.h [hunk 4]
```diff
@@ -108,3 +108,5 @@ class Rectangle : public ConvexPolygon
 
     bool operator==(const Rectangle &p) const;
 };
+
+std::ostream &operator<<(std::ostream &os, const Rectangle &rect);
```
Comment (jonathanlew) [inline src/software/new_geom/rectangle.h]: This should be probably be implemented in `Polygon` and list all the points.

**Review by jonathanlew — COMMENTED** [review]
Looking good, just a minor comment

File: src/software/ai/navigator/path_planner/path_planner_test.cpp (hunk 1 above)
Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: now we use a linear approximation to compute s per metre, and have a constant `PATH_CHECK_INTERVAL_M`

File: src/software/ai/navigator/path_planner/path_planner_test.cpp (hunk 3 above)
Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp:288]: #1207

File: src/software/ai/navigator/path_planner/path_planner_test.cpp [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+#include "software/ai/navigator/path_planner/theta_star_path_planner.h"
+#include "software/new_geom/point.h"
+#include "software/new_geom/rectangle.h"
+#include "software/test_util/test_util.h"
+#include "software/world/world.h"
+
+// interval at which to evaluate the spline value and check
+// if it intersects an obstacle
+static constexpr double PATH_CHECK_INTERVAL = 0.05;
+static constexpr double DEST_CHECK_EPSILON  = 1e-3;
```
[the remark this answers is not in the document]
Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: added

File: src/software/new_geom/rectangle.h (hunk 4 above)
Comment (jordanzliu) [reply src/software/new_geom/rectangle.h]: done

File: src/software/ai/navigator/path_planner/path_planner_test.cpp (hunk 2 above)
Comment (garethellis0) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: But actually, we should probably change "ego" to "a friendly robot", because most people won't get that lol

**Comment by garethellis0** [comment]
Just one small comment, then this is GTG from my end!

File: src/software/ai/navigator/path_planner/path_planner_test.cpp [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+          .start              = Point(0, 0),
+          .dest               = Point(1, 0),
+          .navigable_area     = Rectangle({-1, -1}, {1, 1}),
+          .obstacles          = {},
+          .should_return_path = true},
+
+         {.name           = "Single stationary robot in path",
+          .start          = Point(0, 0),
+          .dest           = Point(2, 0),
+          .navigable_area = Rectangle({-1, -1}, {1, 1}),
```
Comment (MathewMacDougall) [inline src/software/ai/navigator/path_planner/path_planner_test.cpp]: navigable area doesn't contain start and dest

File: src/software/test_util/test_util.cpp [hunk 7]
```diff
@@ -95,4 +95,10 @@ namespace Test
         }
         return game_states;
     }
+
+    Robot TestUtil::createRobotAtPos(const Point &pt)
+    {
+        static RobotId robot_id_counter = 0;
+        return Robot(robot_id_counter++, pt, Vector(), Angle(), Angle(), Timestamp());
```
Comment (MathewMacDougall) [inline src/software/test_util/test_util.cpp]:
```suggestion
        return Robot(robot_id_counter++, pt, Vector(), Angle(), AngularVelocity(), Timestamp());
```
super nit

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/ai/navigator/path_planner/path_planner_test.cpp (hunk 6 above)
Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: fixed

File: src/software/test_util/test_util.cpp (hunk 7 above)
Comment (jordanzliu) [reply src/software/test_util/test_util.cpp]: fixed, also added a comment to this fucntion

**Review by jonathanlew — DISMISSED** [review]
LGTM pending CI :+1:

**Review by akhilveeraghanta — DISMISSED** [review]
:shipit:

**Comment by jordanzliu** [comment]
Looks like the appveyor job is broken?

**Comment by MathewMacDougall** [comment]
@garethellis0 

@jordanzliu in the meantime you should be able to force-push to make it run again.

File: src/software/ai/navigator/path_planner/path_planner_test.cpp [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+{
+    // check for zero length path
+    if (path.valueAt(0.f) == path.valueAt(1.f))
+    {
+        throw std::string("zero length path!");
+    }
+
+    // compute an s value interval that roughly corresponds to length
+    // PATH_CHECK_INTERVAL_M assuming that the path is locally linear
+    double s_per_meter_ratio   = 0.05 / (path.valueAt(0.05) - path.valueAt(0.0)).length();
```
Comment (MathewMacDougall) [inline src/software/ai/navigator/path_planner/path_planner_test.cpp:162]: Should 0.05 be replaced with `PATH_CHECK_INTERVAL_M`?

File: src/software/ai/navigator/path_planner/path_planner_test.cpp [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+    // obstacle otherwise check that the path makes progress toward the destination i.e.
+    // the distance to dest is less at the end of the path than at the start of the path
+    bool dest_in_obstacle = std::any_of(
+        obstacles.begin(), obstacles.end(),
+        [&path](const auto &obs) { return obs.containsPoint(path.valueAt(1.0)); });
+
+    if (dest_in_obstacle &&
+        (path.valueAt(1.0) - dest).length() >= (path.valueAt(0.0) - dest).length())
+    {
+        // fail because no progress to destination
```
Comment (MathewMacDougall) [inline src/software/ai/navigator/path_planner/path_planner_test.cpp]: Bit of a nitpick, but what if we had a testcase where a path started inside an obstacle, and the path also ended inside the same obstacle? The end point of the path would end up outside the obstacle, which would be correct, but the test would fail. I know this is somewhat of an edge case but I can see us writing a test like this.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Really like this overall, just a few nits

File: src/software/ai/navigator/path_planner/path_planner_test.cpp (hunk 8 above)
Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp:162]: 0.05 is a fixed, small s value only used to approximate s/meter, and isn't relevant to `PATH_CHECK_INTERVAL_M`, which is the interval in metres at which to evaluate the spline.

File: src/software/ai/navigator/path_planner/path_planner_test.cpp (hunk 9 above)
Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: we definitely will need a test case where we start inside an obstacle and exit it, which is correct behavior. perhaps we could ignore obstacles that we start in? will explore this more tomorrow

Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: done

File: src/software/ai/navigator/path_planner/path_planner_test.cpp [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    }
+    else if ((path.valueAt(1.0) - dest).length() >= DEST_CHECK_EPSILON_M)
+    {
+        // fail because didn't reach destination
+        std::stringstream fail_ss;
+        fail_ss << "Path ends at " << path.valueAt(1.0) << " but dest is " << dest;
+        throw fail_ss.str();
+    }
+
+    // check that the actual destination of the path
```
Comment (MathewMacDougall) [inline src/software/ai/navigator/path_planner/path_planner_test.cpp]: nit: extra comment

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
just a nit otherwise LGTM

**Comment by garethellis0** [comment]
Just one comment (about renaming "ego") left here from me, then this is GTG from my end.

**Review by jonathanlew — DISMISSED** [review]
Nothing to add beyond what's already been said

File: src/software/ai/navigator/path_planner/path_planner_test.cpp (hunk 2 above)
Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: removed

File: src/software/ai/navigator/path_planner/path_planner_test.cpp (hunk 10 above)
Comment (jordanzliu) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: fixed

**Review by garethellis0 — APPROVED** [review]
LGTM pending other's comments! :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:
